server/src/main/java/io/crate/execution/dml/DynamicIndexer.java

```diff
@@ -0,0 +1,134 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.dml;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.IndexableField;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import io.crate.execution.dml.Indexer.ColumnConstraint;
+import io.crate.execution.dml.Indexer.Synthetic;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.IndexType;
+import io.crate.metadata.Reference;
+import io.crate.metadata.ReferenceIdent;
+import io.crate.metadata.RowGranularity;
+import io.crate.metadata.SimpleReference;
+import io.crate.sql.tree.ColumnPolicy;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.FloatType;
+import io.crate.types.IntegerType;
+import io.crate.types.ShortType;
+import io.crate.types.StorageSupport;
+
+public final class DynamicIndexer implements ValueIndexer<Object> {
+
+    private final ReferenceIdent refIdent;
+    private final Function<ColumnIdent, FieldType> getFieldType;
+    private final Function<ColumnIdent, Reference> getRef;
+    private DataType<?> type = null;
+    private ValueIndexer<Object> indexer;
+
+    public DynamicIndexer(ReferenceIdent refIdent,
+                          Function<ColumnIdent, FieldType> getFieldType,
+                          Function<ColumnIdent, Reference> getRef) {
+        this.refIdent = refIdent;
+        this.getFieldType = getFieldType;
+        this.getRef = getRef;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void indexValue(Object value,
+                           XContentBuilder xcontentBuilder,
+                           Consumer<? super IndexableField> addField,
+                           Consumer<? super Reference> onDynamicColumn,
+                           Map<ColumnIdent, Synthetic> synthetics,
+                           Map<ColumnIdent, ColumnConstraint> toValidate) throws IOException {
+        if (type == null) {
+            type = guessType(value);
+            StorageSupport<?> storageSupport = type.storageSupport();
+            if (storageSupport == null) {
+                throw new IllegalArgumentException(
+                    "Cannot create columns of type " + type.getName() + " dynamically. " +
+                    "Storage is not supported for this type");
+            }
+            boolean nullable = true;
+            int position = -1;
+            Symbol defaultExpression = null;
+            Reference newColumn = new SimpleReference(
+                refIdent,
```

Comment:
`DynamicReference` probably fits better for `DynamicIndexer`?

Reply:
Why? At this point it stops being dynamic

Reply:
OK, I misunderstood the point of `DynamicReference` - I thought we always distinguish that some column was created dynamically. Apparently it's just intermediate thing.